specifyweb/frontend/js_src/lib/components/WorkBench/WbView.tsx

```diff
@@ -49,13 +48,298 @@ import type { WbMapping } from './mapping';
+      )
+      .map(workbench.hot.toVisualColumn);
+
+    if (showResults) {
+      workbench.hot?.updateSettings({ readOnly: true });
+      getHotPlugin(workbench.hot, 'hiddenRows').hideRows(rowsToHide);
+      getHotPlugin(workbench.hot, 'hiddenColumns').hideColumns(colsToHide);
+
+      workbench.utils.toggleCellTypes(
+        'newCells',
+        'remove',
+        spreadsheetContainer.current
+      );
+    } else {
+      getHotPlugin(workbench.hot, 'hiddenRows').showRows(
+        rowsToHide.filter((visualRow) => !initialHiddenRows.includes(visualRow))
+      );
+      getHotPlugin(workbench.hot, 'hiddenColumns').showColumns(
+        colsToHide.filter((visualCol) => !initialHiddenCols.includes(visualCol))
+      );
+    }
+
+    workbench.hot.render();
+  }, [showResults]);
+
+  return (
+    <>
+      <div
+        className="flex items-center justify-between gap-x-1 gap-y-2 whitespace-nowrap"
+        role="toolbar"
+      >
+        <div className="contents">
+          <DataSetName
+            dataset={dataset}
+            getRowCount={() =>
+              hotRef.current?.hotInstance
+                ? dataset.rows.length
+                : hotRef.current!.hotInstance!.countRows() -
+                  hotRef.current!.hotInstance!.countEmptyRows(true)
+            }
+          />
+        </div>
+        <Button.Small
+          aria-haspopup="grid"
+          aria-pressed={showToolkit}
+          onClick={toggleToolkit}
+        >
+          {commonText.tools()}
+        </Button.Small>
+        <span className="-ml-1 flex-1" />
+        {canUpdate || isMapped ? (
+          <Link.Small href={`/specify/workbench/plan/${dataset.id}/`}>
+            {wbPlanText.dataMapper()}
+          </Link.Small>
+        ) : undefined}
+        <WbActionsComponent
+          dataset={dataset}
+          hasUnsavedChanges={hasUnsavedChanges}
+          isUploaded={isUploaded}
+          onRefresh={triggerRefresh}
+          mappings={mappings!}
+          checkDeletedFail={checkDeletedFail}
+          onSpreadsheetUpToDate={spreadsheetUpToDate}
+          workbench={workbench}
+          onToggleResults={toggleResults}
+        />
+      </div>
+      {showToolkit ? (
+        <WbToolkit
+          dataset={dataset}
+          hotRef={hotRef}
+          mappings={mappings!}
+          data={data}
+          handleDatasetDelete={handleDatasetDelete}
+          hasUnsavedChanges={hasUnsavedChanges}
+          triggerRefresh={triggerRefresh}
+        />
+      ) : undefined}
+      <div className="flex flex-1 gap-4 overflow-hidden">
+        <section className="flex-1 overflow-hidden overscroll-none">
+          <WbSpreadsheet
+            dataset={dataset}
+            hotRef={hotRef}
+            isUploaded={isUploaded}
+            data={data}
+            validation={workbench.validation}
+            cells={workbench.cells}
+            disambiguation={workbench.disambiguation}
+            hooks={hooks}
+            mappings={mappings!}
```

Comment:
the `!` in `mappings={mappings!}` might be a mistake
mappings are undefined when dataset does not [yet] have an upload plan
workbench should be able to handle such case (but of course disable all features that require mapping) to let people inspect their dataset/do quick edits before doing the mapping (inspecting the dataset is useful as part of figuring out how the dataset should be mapped)